vision/standard_object/odlc_classify_shape.py

```diff
@@ -0,0 +1,355 @@
+
+    Parameters
+    ----------
+    approx : consts.Contour
+        Approximated version of the contour (polygon approximation with cv2.approxPolyDP())
+
+    Returns
+    -------
+    polygonal_shape : chars.ODLCShape | None
+        Will return one of TRIANGLE, SQUARE, RECTANGLE, TRAPEZOID, PENTAGON, HEXAGON, or OCTAGON
+        from ODLCShape Enum or None if shape does not match any of these.
+    """
+    shape: chars.ODLCShape | None = None
+    match len(approx):
+        case 3:
+            shape = chars.ODLCShape.TRIANGLE
+        case 4:
+            angles: npt.NDArray[npt.Shape["*"], npt.Float64] = get_angles(approx)
+            if np.all(compare_angles(angles, 90, QUAD_ANG_THRESH)):
+                lengths: npt.NDArray[npt.Shape["*"], npt.Float64] = get_lengths(approx)
+                if (lengths[0]) / (np.sum(lengths) / 4) < QUAD_LEN_THRESH:
+                    shape = chars.ODLCShape.SQUARE
+                else:
+                    shape = chars.ODLCShape.RECTANGLE
+            else:
+                shape = chars.ODLCShape.TRAPEZOID
+        case 5:
+            shape = chars.ODLCShape.PENTAGON
+        case 6:
+            shape = chars.ODLCShape.HEXAGON
+        case 7:
+            shape = chars.ODLCShape.HEPTAGON
+        case 8:
+            shape = chars.ODLCShape.OCTAGON
+        case _:
+            shape = None
+
+    return shape
+
+
+def classify_circular(contour: consts.Contour) -> chars.ODLCShape:
+    """
+    Determines which circular shape the given contour is (assumed to be a circular shape)
+
+    Parameters
+    ----------
+    contour : consts.Contour
+        The original contour to evaluate
+
+    Returns
+    -------
+    circular_shape : chars.ODLCShape
+        Will return one of CIRCLE, SEMICIRCLE, QUARTER_CIRCLE from ODLCShape Enum
+
+    Notes
+    -----
+    Alternative method to determine which circular shape could be to see how many right angles
+    the given shape has a quarter-circle may have 1 or 3 (if corner between straight line and
+    curve show up as a right angle), and a semicircle would have 0 or 2 for same reason. But
+    a regular circle would have 0 right angles, which may conflict with semicircle.
+    """
+    approx: consts.Contour = cv2.approxPolyDP(contour, cv2.arcLength(contour, True) * 0.01, True)
+    max_dist: npt.Float64 = np.max(get_lengths(approx))
+    perimeter: npt.Float64 = cv2.arcLength(contour, True)
+
+    if abs((max_dist / perimeter) - QUARTER_CIRCLE_RATIO) < CIRCULAR_RATIO_RANGE:
+        return chars.ODLCShape.QUARTER_CIRCLE
+
+    if abs((max_dist / perimeter) - SEMICIRCLE_RATIO) < CIRCULAR_RATIO_RANGE:
+        return chars.ODLCShape.SEMICIRCLE
+    return chars.ODLCShape.CIRCLE
+
+
+def classify_shape(
+    contours: list[consts.Contour],
+    hierarchy: consts.Hierarchy,
+    index: int,
+    image_dims: tuple[int, int],
+    approx_contour: consts.Contour | None = None,
+) -> chars.ODLCShape | None:
```

Comment:
Add a function that takes in the contours and hierarchy and loops through them all and returns an array of `BoundingBoxes` with the shape as an object attribute